staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_discovery_controller.go

```diff
@@ -0,0 +1,209 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiserver
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/endpoints/discovery"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"k8s.io/kube-apiextensions-server/pkg/apis/apiextensions"
+	informers "k8s.io/kube-apiextensions-server/pkg/client/informers/internalversion/apiextensions/internalversion"
+	listers "k8s.io/kube-apiextensions-server/pkg/client/listers/apiextensions/internalversion"
+)
+
+type CustomResourceDiscoveryController struct {
+	versionHandler *customResourceVersionDiscoveryHandler
+	groupHandler   *customResourceGroupDiscoveryHandler
+
+	customResourceLister  listers.CustomResourceLister
+	customResourcesSynced cache.InformerSynced
+
+	// To allow injection for testing.
+	syncFn func(version schema.GroupVersion) error
+
+	queue workqueue.RateLimitingInterface
+}
+
+func NewCustomResourceDiscoveryController(customResourceInformer informers.CustomResourceInformer, versionHandler *customResourceVersionDiscoveryHandler, groupHandler *customResourceGroupDiscoveryHandler) *CustomResourceDiscoveryController {
+	c := &CustomResourceDiscoveryController{
+		versionHandler:        versionHandler,
+		groupHandler:          groupHandler,
+		customResourceLister:  customResourceInformer.Lister(),
+		customResourcesSynced: customResourceInformer.Informer().HasSynced,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "CustomResourceDiscoveryController"),
+	}
+
+	customResourceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addCustomResource,
+		UpdateFunc: c.updateCustomResource,
+		DeleteFunc: c.deleteCustomResource,
+	})
+
+	c.syncFn = c.sync
+
+	return c
+}
+
+func (c *CustomResourceDiscoveryController) sync(version schema.GroupVersion) error {
+
+	foundVersion := false
+	foundGroup := false
+
+	apiVersionsForDiscovery := []metav1.GroupVersionForDiscovery{}
+	apiResourcesForDiscovery := []metav1.APIResource{}
+
+	customResources, err := c.customResourceLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	for _, customResource := range customResources {
+		if customResource.Spec.Group != version.Group {
+			continue
+		}
+		foundGroup = true
+		apiVersionsForDiscovery = append(apiVersionsForDiscovery, metav1.GroupVersionForDiscovery{
+			GroupVersion: customResource.Spec.Group + "/" + customResource.Spec.Version,
+			Version:      customResource.Spec.Version,
+		})
+
+		if customResource.Spec.Version != version.Version {
+			continue
+		}
+		foundVersion = true
+
+		apiResourcesForDiscovery = append(apiResourcesForDiscovery, metav1.APIResource{
```

Comment:
don't we need a check here that the resources are accepted? If that's just not implemented, please add a TODO.